ddtrace/tracer/sqlcomment.go

```diff
@@ -0,0 +1,135 @@
+	SQLInjectionModeFull SQLCommentInjectionMode = 2
+)
+
+// Key names for SQL comment tags.
+const (
+	sqlCommentKeySamplingPriority = "ddsp"
+	sqlCommentTraceID             = "ddtid"
+	sqlCommentSpanID              = "ddsid"
+	sqlCommentService             = "ddsn"
+	sqlCommentVersion             = "ddsv"
+	sqlCommentEnv                 = "dde"
+)
+
+// SQLCommentCarrier is a carrier implementation that injects a span context in a SQL query in the form
+// of a sqlcommenter formatted comment prepended to the original query text.
+// See https://google.github.io/sqlcommenter/spec/ for more details.
+type SQLCommentCarrier struct {
+	Query  string
+	Mode   SQLCommentInjectionMode
+	SpanID uint64
+}
+
+// Inject injects a span context in the carrier's query.
+func (c *SQLCommentCarrier) Inject(spanCtx ddtrace.SpanContext) error {
+	c.SpanID = random.Uint64()
+	tags := make(map[string]string)
+	switch c.Mode {
+	case SQLInjectionDisabled:
+		return nil
+	case SQLInjectionModeFull:
+		samplingPriority := 0
+		var traceID uint64
+		ctx, ok := spanCtx.(*spanContext)
+		if ok {
+			if sp, ok := ctx.samplingPriority(); ok {
+				samplingPriority = sp
+			}
+			if ctx.TraceID() > 0 {
+				traceID = ctx.TraceID()
+			}
+		}
+		if traceID == 0 {
+			traceID = c.SpanID
+		}
+		tags[sqlCommentTraceID] = strconv.FormatUint(traceID, 10)
+		tags[sqlCommentSpanID] = strconv.FormatUint(c.SpanID, 10)
+		tags[sqlCommentKeySamplingPriority] = strconv.Itoa(samplingPriority)
+		fallthrough
+	case SQLInjectionModeService:
+		ctx, ok := spanCtx.(*spanContext)
+		var env, version string
+		if ok {
+			if e, ok := ctx.meta(ext.Environment); ok {
+				env = e
+			}
+			if v, ok := ctx.meta(ext.Version); ok {
+				version = v
+			}
+		}
+		if globalconfig.ServiceName() != "" {
+			tags[sqlCommentService] = globalconfig.ServiceName()
+		}
+		if env != "" {
+			tags[sqlCommentEnv] = env
+		}
+		if version != "" {
+			tags[sqlCommentVersion] = version
+		}
+	}
+	c.Query = commentQuery(c.Query, tags)
+	return nil
+}
+
+// commentQuery returns the given query with the tags from the SQLCommentCarrier applied to it as a
+// prepended SQL comment. The format of the comment follows the sqlcommenter spec.
+// See https://google.github.io/sqlcommenter/spec/ for more details.
+func commentQuery(query string, tags map[string]string) string {
+	if len(tags) == 0 {
+		return ""
+	}
+	serializedTags := make([]string, 0, len(tags))
+	for k, v := range tags {
+		eKey := url.QueryEscape(k)
+		eKey = strings.Replace(eKey, "+", "%20", -1)
+		sKey := strings.ReplaceAll(eKey, "'", "\\'")
+		eVal := url.QueryEscape(v)
+		eVal = strings.Replace(eVal, "+", "%20", -1)
+		escVal := strings.ReplaceAll(eVal, "'", "\\'")
+		sValue := fmt.Sprintf("'%s'", escVal)
+		serializedTags = append(serializedTags, fmt.Sprintf("%s=%s", sKey, sValue))
+	}
+	sort.Strings(serializedTags)
```

Comment:
Why do they need to be sorted? There is a lot of overhead introduced in this function. Can we avoid `fmt.Sprintf` throughout and use `strings.Builder` instead to construct the entire final string?

Reply:
That's [part of the spec](https://google.github.io/sqlcommenter/spec/#sorting). We can decide to diverge more from the spec but I think the default should be to try and be compliant. 

Regarding the `strings.Builder`, yes. That doesn't affect our compliance and should make things faster and with less overhead.

Reply:
Just thought of something you might like regarding sorting: since we have a fixed set of keys that we're injecting, we can use a slice of key values and control how we add them so we can guarantee they're sorted without actually calling a `sort`. 

It might make the `Inject` a bit more complex because it would need to consider adding tags based on the alphanumeric place of the keys but it would achieve the thing you want: skipping the sort operation while still guaranteeing that the output is sorted.

Reply:
I updated this to define a slice of `orderedKeys` to guarantee sorted output without sorting. This might be a little brittle for when we add new keys but the testing should help (not adding the new key to the orderedKeys would result in that new key not being included in comments.